scripts/zones/Sacrarium/npcs/Small_Keyhole.lua

```diff
@@ -3,35 +3,42 @@
 -- Area: Sacrarium
 -- !pos 99.772 -1.614 51.545 28
 -----------------------------------
-package.loaded["scripts/zones/Sacrarium/TextIDs"] = nil;
+package.loaded["scripts/zones/Sacrarium/TextIDs"] = nil
 -----------------------------------
-require("scripts/zones/Sacrarium/TextIDs");
+require("scripts/zones/Sacrarium/TextIDs")
+require("scripts/zones/Sacrarium/MobIDs")
+require("scripts/globals/keyitems")
+require("scripts/globals/npc_util")
 -----------------------------------
 
 function onTrigger(player,npc)
-
-    local SmallKeyholeID = npc:getID();
-    local DoorID = GetNPCByID(SmallKeyholeID):getID() - 3;
-
-    if (player:hasKeyItem(dsp.ki.TEMPLE_KNIGHT_KEY)) then
-        GetNPCByID(DoorID):openDoor(15);
+    if player:hasKeyItem(dsp.ki.TEMPLE_KNIGHT_KEY) then
+        GetNPCByID(npc:getID() - 3):openDoor(15)
     else
-        player:messageSpecial(SMALL_KEYHOLE);
+        player:messageSpecial(SMALL_KEYHOLE)
     end
-end;
+end
 
 function onTrade(player,npc,trade)
-
-    if (trade:hasItemQty(1659,1) and trade:getItemCount() == 1) then
-        player:messageSpecial(CORAL_KEY_TRADE);
-        SetServerVariable("SACRARIUM_Coral_Key_trade",os.time());
-        player:tradeComplete();
-        -- print(os.time());
+    if npcUtil.tradeHas(trade, 1659) then
+        if npc:getLocalVar("canTradeSecondKey") == 0 then
+            npc:setLocalVar("canTradeSecondKey", 1)
+            player:startEvent(100)
+        else
+            player:messageSpecial(CANNOT_TRADE_NOW)
+        end
     end
-end;
+end
 
 function onEventUpdate(player,csid,option)
+    if csid == 100 and option == 2 then
+        GetNPCByID(SMALL_KEYHOLE):setLocalVar("canTradeSecondKey", 0)
+    end
 end
 
 function onEventFinish(player,csid,option)
-end;
+    if csid == 100 then
```

Comment:
shouldn't we also have an option code checked here? this will break the key always, even if the user gets kicked out of the event. (I'm genuinely uncertain of the retail behavior, been too many years)

Reply:
unsure re: retail. the event is about 2.5 seconds long, and supposed to break the coral key regardless of what happens with the other sublock.

Reply:
does the event let you escape key out?

Reply:
Escape key does nothing during event.

Reply:
k we're good then